src/lib/sw_engine/tvgSwRaster.cpp

```diff
@@ -573,6 +573,66 @@ static bool _rasterImage(SwSurface* surface, uint32_t *img, uint32_t w, TVG_UNUS
     return true;
 }
 
+static uint32_t _applyBilinearInterpolation(const uint32_t *img, uint32_t w, uint32_t h, uint32_t rX, uint32_t rY, float fX, float fY)
+{
+    auto dX = fX - rX;
```

Comment:
and what do you think of using a floor() function instead of passing rX and rY?
I didn't check the code below, since I wanted to answer fast, but now I will :)

```
auto rX = static_cast<uint32_t>(floor(fX));
auto rY = static_cast<uint32_t>(floor(fY));

auto dX = fX - rX;
auto dY = fY - rY;

//0.0039f is a condition for the interpolation ratio to be at least 1.
if (abs(dX) > 0.0039f && abs(dY) > 0.0039f) {
        float w1 = dX;
        float h1 = dY;
        uint32_t p1, p2, p3, p4;

        p1 = img[rX + rY  * w];
        p2 = img[rX + 1 + rY * w];
        p3 = img[rX + 1 + (rY + 1) * w];
        p4 = img[rX + (rY + 1) * w];

        if (p1 == p2 && p1 == p3 && p1 == p4) {
            return img[rX + (rY * w)];
        } else {
            auto c1 = COLOR_INTERPOLATE(p1, static_cast<uint32_t>((1 - h1) * 255), p4, static_cast<uint32_t>(h1 * 255));
            auto c2 = COLOR_INTERPOLATE(p2, static_cast<uint32_t>((1 - h1) * 255), p3, static_cast<uint32_t>(h1 * 255));
            return COLOR_INTERPOLATE(c1, static_cast<uint32_t>((1 - w1) * 255), c2, static_cast<uint32_t>(w1 * 255));
        }
}
return img[rX + (rY * w)];
```

Reply:
what I wrote seems to be ok. 
I named the points as follows:
p1 -- p2
 |        |
p4 -- p3

Reply:
I like the above. I suggest removing w1 and h1 (use dX and dY directly).
In COLOR_INTERPOLATE I am a little suspicious that a1 + a2 won't sum up 255 (but 254). Maybe we can introduce function like:
```
static inline uint32_t COLOR_INTERPOLATE_2(uint32_t c1, uint32_t c2, uint8_t a)
{
    return COLOR_INTERPOLATE(c1, a, c2, 255-a);
}
```
It will reduce one float multiplication and casting and ensure a1+a2=255. What do you think?

Reply:
Maybe we can also drop floor and use just `static_cast<uint32_t>(fX)`.
`floor` may be slow (we call it multiple times) and not vectorize as SIMD.

Reply:
@mgrudzinska @mmaciola 
Based on review comments, I've modified the code to be simpler. As a result, unnecessary if-checks and calculations are reduced. Thank you for review.